src/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateManager.ProjectStates.cs

```diff
@@ -100,17 +92,37 @@ private ProjectAnalyzerStateSets CreateProjectStateSets(Project project)
             /// </summary>
             private ProjectAnalyzerStateSets UpdateProjectStateSets(Project project)
             {
-                var projectStateSets = CreateProjectStateSets(project);
+                ProjectAnalyzerReferenceChangedEventArgs? analyzerReferenceChangedEventArgs = null;
+                ProjectAnalyzerStateSets projectStateSets;
+
+                // This code is called concurrently for a project, so the lock prevents duplicated effort calculating StateSets.
+                lock (_projectAnalyzerStateMapLock)
+                {
+                    var oldProjectStateSets = TryGetProjectStateSets(project);
+
+                    if (oldProjectStateSets != null)
+                    {
+                        // Someone calculated the value while we waited on the lock, we can use that
+                        projectStateSets = oldProjectStateSets.Value;
+                    }
+                    else
+                    {
+                        projectStateSets = CreateProjectStateSets(project);
 
-                RaiseProjectAnalyzerReferenceChangedIfNeeded(project, projectStateSets.MapPerReferences, projectStateSets.StateSetMap);
+                        analyzerReferenceChangedEventArgs = GetProjectAnalyzerReferenceChangedEventArgs(project, projectStateSets.MapPerReferences, projectStateSets.StateSetMap);
+
+                        // update cache. 
+                        _projectAnalyzerStateMap = _projectAnalyzerStateMap.SetItem(project.Id, projectStateSets);
+                    }
+                }
 
-                // update cache. 
-                _projectAnalyzerStateMap[project.Id] = projectStateSets;
+                if (analyzerReferenceChangedEventArgs != null)
+                    RaiseProjectAnalyzerReferenceChanged(analyzerReferenceChangedEventArgs);
```

Comment:
>RaiseProjectAnalyzerReferenceChanged [](http://example.com/codeflow?start=20&length=36)

Re-jiggered this code to do the calculation of the eventargs inside the lock and the notification outside.

Reply:
wonderful.